nilearn/connectome/rena_clustering.py

```diff
@@ -0,0 +1,646 @@
+
+    n_labels : int
+        The number of clusters in the current iteration.
+
+    connectivity : a sparse matrix in COOrdinate format.
+
+    masked_data : array like
+        2D data matrix.
+
+    threshold : float in the close interval [0, 1]
+        The treshold is setted to handle eccentricities.
+        In practice it is 1e-7.
+
+    Returns
+    -------
+    reduced_connectivity : a sparse matrix in COOrdinate format.
+
+    reduced_masked_data: array like
+        2D data matrix.
+    """
+    n_features = len(labels)
+
+    incidence = coo_matrix(
+        (np.ones(n_features), (labels, np.arange(n_features))),
+        shape=(n_labels, n_features), dtype=np.float32).tocsc()
+
+    inv_sum_col = dia_matrix(
+        (np.array(1. / incidence.sum(axis=1)).squeeze(), 0),
+        shape=(n_labels, n_labels))
+
+    incidence = inv_sum_col * incidence
+
+    reduced_masked_data = (incidence * masked_data.T).T
+    reduced_connectivity = (incidence * connectivity) * incidence.T
+
+    reduced_connectivity = reduced_connectivity - dia_matrix(
+        (reduced_connectivity.diagonal(), 0),
+        shape=(reduced_connectivity.shape))
+
+    i_idx, j_idx = reduced_connectivity.nonzero()
+
+    weights_ = np.sum(
+        (reduced_masked_data[:, i_idx] - reduced_masked_data[:, j_idx]) ** 2,
+        axis=0)
+    weights_ = np.maximum(threshold, weights_)
+    reduced_connectivity.data = weights_
+
+    return reduced_connectivity, reduced_masked_data
+
+
+def nearest_neighbor_grouping(connectivity, masked_data, n_clusters,
+                              threshold):
+    """Cluster using nearest agglomeration: merge clusters according to their
+    nearest neighbors, then the data and the connectivity are reduced.
+
+    Parameters
+    ----------
+    connectivity : a sparse matrix in COOrdinate format.
+        Weighted connectivity matrix
+
+    masked_data : numpy array of shape [n_samples, n_features]
+        Image in brain space transformed into 2D data matrix
+
+    n_clusters : int
+        The number of clusters to find.
+
+    threshold : float in the close interval [0, 1]
+        The treshold is setted to handle eccentricities.
+        In practice it is 1e-7.
+
+    Returns
+    -------
+    reduced_connectivity : a sparse matrix in COOrdinate format.
+
+    reduced_masked_data :  array like
+        2D data matrix.
+
+    labels : array like
+        It contains the clusters assignation.
+    """
+    # Nearest neighbor conenctivity
+    nn_connectivity = _nn_connectivity(connectivity, threshold)
+    n_features = connectivity.shape[0]
+
+    n_labels = n_features - (nn_connectivity + nn_connectivity.T).nnz / 2
+
+    if n_labels < n_clusters:
+        # remove edges so that the final number of clusters is not less than
+        # n_clusters (to achieve the desired number of clusters)
+        n_edges = n_features - n_clusters
+
+        nn_connectivity = (nn_connectivity + nn_connectivity.T)
+
```

Comment:
Maybe that's a matter of taste, but i don't like skipping lines without reason. I do it after an indented block or whenever pep8 requires, but not here.